### ISSUE TITLE
nan's after some epochs

### ISSUE THREAD

**Issue by gaberoo** [RasmussenLab/vamb#12]
Something is going wrong...

`python3 ~/Projects/vamb/run.py output ../final.contigs.fa ../minimap2/*.bam -m 2000 -p 8`

I'm on macOS 10.14.

log.txt
```
Starting Vamb version 1.0.1
        Date and time is 2019-03-10 08:43:14.611644

Calculating TNF
        Processed 354671017 bases in 73322 sequences.
        Calculated TNF in 49.32 seconds.

Calculating RPKM
        Parsing 192 BAM files with 8 subprocesses.
...
        Calculated RPKM in 1052.27 seconds.

Creating and training VAE
        Created VAE
        Created dataloader and mask
        Number of sequences unsuitable for encoding: 0
        Number of sequences remaining: 73322
        Network properties:
        CUDA: False
        Alpha: 0.05
        Beta: 200.0
        Dropout: 0.2
        N hidden: 325, 325
        N latent: 40

        Training properties:
        N epochs: 500
        Starting batch size: 64
        Batchsteps: 25, 75, 150, 300
        Learning rate: 0.001
        N sequences: 73322
        N samples: 192

        Epoch: 1        Loss: 0.807687  CE: 4.2947513   SSE: 67.509104  KLD: 54.6405    Batchsize: 64
        Epoch: 2        Loss: 0.747882  CE: 3.9706422   SSE: 59.752448  KLD: 67.5331    Batchsize: 64
        Epoch: 3        Loss: 0.732748  CE: 3.8860711   SSE: 59.077767  KLD: 70.6974    Batchsize: 64
        Epoch: 4        Loss: 0.723977  CE: 3.8379009   SSE: 58.487195  KLD: 71.8985    Batchsize: 64
        Epoch: 5        Loss: 0.718393  CE: 3.8071205   SSE: 58.094642  KLD: 72.8744    Batchsize: 64
        Epoch: 6        Loss: 0.714395  CE: 3.7855409   SSE: 57.548828  KLD: 73.6923    Batchsize: 64
        Epoch: 7        Loss: 0.711017  CE: 3.7669103   SSE: 57.338126  KLD: 74.2194    Batchsize: 64
        Epoch: 8        Loss: 0.708550  CE: 3.7537999   SSE: 57.076424  KLD: 74.2049    Batchsize: 64
        Epoch: 9        Loss: 0.706744  CE: 3.7436445   SSE: 56.975041  KLD: 74.7307    Batchsize: 64
        Epoch: 10       Loss: 0.705119  CE: 3.7345498   SSE: 56.923979  KLD: 75.0273    Batchsize: 64
        Epoch: 11       Loss: 0.703744  CE: 3.7271899   SSE: 56.752982  KLD: 75.1727    Batchsize: 64
        Epoch: 12       Loss: 0.702570  CE: 3.7205976   SSE: 56.617877  KLD: 75.7071    Batchsize: 64
        Epoch: 13       Loss: 0.701273  CE: 3.7137799   SSE: 56.575563  KLD: 75.3124    Batchsize: 64
        Epoch: 14       Loss: 0.700491  CE: 3.7095677   SSE: 56.500341  KLD: 75.3655    Batchsize: 64
        Epoch: 15       Loss: 0.699735  CE: 3.7054268   SSE: 56.384222  KLD: 75.6489    Batchsize: 64
        Epoch: 16       Loss: 0.698815  CE: 3.7007463   SSE: 56.187604  KLD: 75.6258    Batchsize: 64
        Epoch: 17       Loss: 0.698197  CE: 3.6969053   SSE: 56.292795  KLD: 75.9304    Batchsize: 64
....
        Epoch: 55       Loss: 0.684398  CE: 3.6239755   SSE: 53.925301  KLD: 77.9252    Batchsize: 128
        Epoch: 56       Loss: 0.684124  CE: 3.6226431   SSE: 53.834398  KLD: 77.9213    Batchsize: 128
        Epoch: 57       Loss: 0.684131  CE: 3.6224621   SSE: 53.915936  KLD: 78.0038    Batchsize: 128
        Epoch: 58       Loss: 0.684008  CE: 3.6216544   SSE: 53.905198  KLD: 78.2170    Batchsize: 128
        Epoch: 59       Loss: 0.684033  CE: 3.6218417   SSE: 53.907004  KLD: 78.1390    Batchsize: 128
        Epoch: 60       Loss: 0.683866  CE: 3.6208075   SSE: 53.864307  KLD: 78.4251    Batchsize: 128
        Epoch: 61       Loss: 0.683701  CE: 3.6201069   SSE: 53.847837  KLD: 78.1658    Batchsize: 128
        Epoch: 62       Loss: 0.683533  CE: 3.6192230   SSE: 53.814829  KLD: 78.1990    Batchsize: 128
        Epoch: 63       Loss: 0.683592  CE: 3.6194902   SSE: 53.775092  KLD: 78.3995    Batchsize: 128
        Epoch: 64       Loss: 0.683298  CE: 3.6182224   SSE: 53.759926  KLD: 77.9298    Batchsize: 128
        Epoch: 65       Loss: 0.683324  CE: 3.6179563   SSE: 53.754940  KLD: 78.5331    Batchsize: 128
        Epoch: 66       Loss: 0.683060  CE: 3.6167170   SSE: 53.769388  KLD: 78.1741    Batchsize: 128
        Epoch: 67       Loss: 0.683236  CE: 3.6171935   SSE: 53.776952  KLD: 78.8666    Batchsize: 128
        Epoch: 68       Loss: 0.683023  CE: 3.6162189   SSE: 53.811286  KLD: 78.4733    Batchsize: 128
        Epoch: 69       Loss: 0.682970  CE: 3.6157823   SSE: 53.814611  KLD: 78.6723    Batchsize: 128
        Epoch: 70       Loss: 0.682921  CE: 3.6157700   SSE: 53.754230  KLD: 78.4723    Batchsize: 128
        Epoch: 71       Loss: 0.682810  CE: 3.6150326   SSE: 53.817880  KLD: 78.4649    Batchsize: 128
...
        Epoch: 272      Loss: 0.673535  CE: 3.5656217   SSE: 51.935407  KLD: 81.2284    Batchsize: 512
        Epoch: 273      Loss: 0.673623  CE: 3.5658599   SSE: 51.966956  KLD: 81.4935    Batchsize: 512
        Epoch: 274      Loss: 0.673627  CE: 3.5659335   SSE: 51.968505  KLD: 81.4161    Batchsize: 512
        Epoch: 275      Loss: 0.673581  CE: 3.5658905   SSE: 51.963353  KLD: 81.1226    Batchsize: 512
        Epoch: 276      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 277      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 278      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 279      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 280      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 281      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 282      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
        Epoch: 283      Loss: nan       CE: nan SSE: nan        KLD: nan        Batchsize: 512
```

**Comment by jakobnissen** [comment]
Thanks for opening an issue about this bug in Vamb!

Hm, it seems like it's producing NANs in the training process. My bet is that your VAE learns that the depth of a contig i a sample is zero, and produces a probability of exactly 0.0, which crashes the cross entropy calculation because you can't take the log of it.

One possible workaround is to add a very small pseudocount (e.g. 1e-25) to each of the estimated probabilities. But perhaps there's some nice numerically stable function in PyTorch already. I'll look at it tomorrow or later tonight, and fix it in the master branch.

**Comment by gaberoo** [comment]
If it's just in the calculation of the entropy, then there is an easy trick:

H = - Sum[ p * log(p) ]

There's a problem that 0.0*log(0.0) = -inf or nan in code. However, mathematically 0*log(0) is zero (if you take the limits. So the trick is to do (e.g. in some kind of Pythonized code):
```
H = - np.sum([ ( p*log(p) if p > 0.0 else 0.0 ) for p in probs ])
```

**Comment by jakobnissen** [comment]
Dear @gaberoo 

While it is true that x\*log(x) converges to zero as x does, this is not true in the general case of y\*log(x). I.e. if the true value (y) is nonzero and the network erroneously produces a zero (x), then the network should be penalized with a high (but finite) error.
Amazingly, PyTorch is able to differentiate across these kind of filtering operations.

I have decided to address this by adding a pseudocount of 1e-9 to the predicted depth after softmaxing. However, before pushing to `master` (or `dev` branch), I will need to run some tests to make sure it doesn't horrifically break Vamb.

If you want to get going immediately and not way a day or two for the fix, simply change the code in your own copy of Vamb. I'm planning to change the code in [`encode.py` line 229](https://github.com/jakobnissen/vamb/blob/master/encode.py#L229) from:
`ce = - (depths_out.log() * depths_in).sum(dim=1).mean()`
to:
`ce = - ((depths_out + 1e-9).log() * depths_in).sum(dim=1).mean()`

If you do decide this, please let me know if it fixed the issue or if I need to look for any other numerical instabilities!

**Comment by gaberoo** [comment]
Great, thanks! I'll look into it.

**Comment by gaberoo** [comment]
Looks good so far.

```
        Epoch: 1        Loss: 0.808921  CE: 4.3022138   SSE: 67.398104  KLD: 54.0473    Batchsize: 64
        Epoch: 2        Loss: 0.749711  CE: 3.9803042   SSE: 60.022812  KLD: 67.4002    Batchsize: 64
        Epoch: 3        Loss: 0.732901  CE: 3.8865031   SSE: 59.152954  KLD: 71.0731    Batchsize: 64
        Epoch: 4        Loss: 0.723607  CE: 3.8354157   SSE: 58.478170  KLD: 72.5587    Batchsize: 64
        Epoch: 5        Loss: 0.718176  CE: 3.8057138   SSE: 58.034014  KLD: 73.3489    Batchsize: 64
        Epoch: 6        Loss: 0.714331  CE: 3.7851134   SSE: 57.619609  KLD: 73.5866    Batchsize: 64
        Epoch: 7        Loss: 0.711443  CE: 3.7698529   SSE: 57.172702  KLD: 73.8569    Batchsize: 64
        Epoch: 8        Loss: 0.709022  CE: 3.7566102   SSE: 57.088032  KLD: 73.8847    Batchsize: 64
        Epoch: 9        Loss: 0.707278  CE: 3.7469653   SSE: 56.894618  KLD: 74.4449    Batchsize: 64
...
        Epoch: 315      Loss: 0.672333  CE: 3.5602202   SSE: 51.531626  KLD: 80.6063    Batchsize: 1024
        Epoch: 316      Loss: 0.672216  CE: 3.5595827   SSE: 51.557299  KLD: 80.5188    Batchsize: 1024
        Epoch: 317      Loss: 0.672328  CE: 3.5601543   SSE: 51.550001  KLD: 80.6046    Batchsize: 1024
        Epoch: 318      Loss: 0.672282  CE: 3.5598616   SSE: 51.543157  KLD: 80.6813    Batchsize: 1024
        Epoch: 319      Loss: 0.672383  CE: 3.5602341   SSE: 51.580201  KLD: 80.8441    Batchsize: 1024
        Epoch: 320      Loss: 0.672454  CE: 3.5608848   SSE: 51.528331  KLD: 80.6211    Batchsize: 1024
	Epoch: 321	Loss: 0.672414	CE: 3.5603594	SSE: 51.616521	KLD: 80.8016	Batchsize: 1024
	Epoch: 322	Loss: 0.672462	CE: 3.5608766	SSE: 51.553784	KLD: 80.6199	Batchsize: 1024
```

Should I be worried that the KL divergence has more or less steadily increased since the start? I do not see an increase followed by a decrease (yet, at 330/500 epochs).

**Comment by jakobnissen** [comment]
@gaberoo 

Great! I'll add this fix to the `dev` branch.
Regarding the increasing KL-divergence, we've seen different behaviours with different datasets. Vamb seems to do well either way.
Honestly, I have no idea what that implies. Sure, it means that the network keeps finding more accurate encodings/decodings, and this gain in accuracy more than makes up for the increase in KL. The lack of decrease in KL also means that the network is probably not overfitting too much. But is it good or bad? I can't really tell.

I'm closing this issue since the original problem seem to be resolved. As always, you're welcome to open another one (or reopen this, if the problem rears its head again).

**Comment by jakobnissen** [comment]
Reopened this issue because I have been experiencing it in a test run. For future reference, the test run was with a modified version of Vamb using only RPKM and not TNF. Interestingly, the `nan`s appear right at a batchsize increase.
I'm utterly stumped as to why this happens.

```
        Epoch: 299      Loss: 0.698978  CE: 3.8315043   KLD: 94.6385    Batchsize: 512
        Epoch: 300      Loss: 0.698972  CE: 3.8314074   KLD: 94.7266    Batchsize: 512
        Epoch: 301      Loss: nan       CE: nan KLD: nan        Batchsize: 1024
        Epoch: 302      Loss: nan       CE: nan KLD: nan        Batchsize: 1024
```

**Comment by gaberoo** [comment]
I added the pseudo count and that fixed the error for me...

**Comment by jakobnissen** [comment]
Possibly due to [how logsigma is calculated](https://github.com/RasmussenLab/vamb/blob/9c9af8c6a734c1183bc9d472b95d10a8fc0acdb6/vamb/encode.py#L213).
Softplus retricts logsigma to be a positive number, but the optimizer pushes towards some negative number. Result could be float underflow in the layer before, as the optimizer pushes its bias towards negative infinity.
I will remove the softplus activation layer and see how that impacts performance. Closing this issue until the bug raises its head again.
